lib/bitcoinerlab.js

```
@@ -0,0 +1,144 @@
+const secp256k1 = require('@exodus/bitcoinerlab-secp256k1')
+let elliptic
+
+const verbose = false
+function loadLegacy (method, message) {
+  if (verbose) {
+    if (!message) {
+      console.warn(`${method} is not implemented, using legacy approach...`)
+    } else console.warn(`${method} | ${message}`)
+  }
+  if (!elliptic) {
+    elliptic = require('./elliptic')
+  }
+}
+
+function compressed (array) {
+  return array.length === 33
+}
+
+module.exports = {
+  contextRandomize: function () {
+    return 0
+  },
+  privateKeyVerify: function (privateKey) {
+    return secp256k1.isPrivate(privateKey) ? 0 : 1
+  },
+  privateKeyNegate: function (privateKey) {
+    try {
+      privateKey.set(secp256k1.privateNegate(privateKey))
+    } catch (e) {
```

This catches all errors. Are we sure not other errors are possible? Isn't there a better way to detect the method isn't implemented? Or at least verify that we have a 'not implemented' error in the catch block?

---

Unlike legacy `secp256k1`, `bitcoinerlab` fails on `negate overflowed private key` and `negate N private key`

---

OK, not sure that answers my question(s).

---

Will try to find out what errors were don't remember it rn

---

>Are we sure no other errors are possible?

The only error `privateNegate` throws is `Expected Private`

Not sure I understand other questions

---

So the `Expected Private` error means (I'm guessing) that we did not supply a (correct) `privateKey`. But we treat it as if `secp256k1.privateNegate` is not implemented. Am I misunderstanding something?

---

`bitcoinerlab.privateNegate` doesn't consider these as private keys and it throws.
`util.BN_ZERO.toArrayLike(Buffer, 'be', 32)`,
`util.ec.curve.n.toArrayLike(Buffer, 'be', 32)`
`util.ec.curve.n.addn(10).toArrayLike(Buffer, 'be', 32)`

Therefore I added `elliptic.privateKeyNegate` as a fallback

---

If it errors here on invalid input private keys (overflowed private key & N private key are both not safe), then I think that's probably more sane behavior.

`secp256k1.privateNegate(privateKey)` can return null, please throw on that rather than setting `privateKey`  to null.

There is difference in behavior here though related to the above issues.

1. [elliptic seems to be wrapping around N](https://github.com/ExodusMovement/secp256k1-node/blob/master/lib/elliptic.js#L81)
2. [while bitcoinerlab seems to return null if it overflows](https://github.com/bitcoinerlab/secp256k1/blob/main/index.js#L129-L130)

---

Ah, I see. Understood.

---

The correct behavior is to return an error if the input key is not a valid private key. Private keys are only valid if they are in the range $[1, n-1]$ (inclusive). `@noble/secp256k1`'s behavior is correct here. 

I don't see how that could be abused in the context of private key negation. But by falling back to elliptic's weird behavior which accepts invalid keys, we're deviating from expected norms here.